foundation/gateway/pkg/adapter/kubernetes/adapter.go

```diff
@@ -141,11 +176,121 @@ func (a *Adapter[T]) Load(ctx context.Context, obj *T) error {
 		return fmt.Errorf("%w: failed to convert %s: %w", model.ErrValidation, a.gvr.Resource, err)
 	}
 	*obj = converted
 	return nil
 }
+
+// Create implements the port.WriterRepo interface.
+func (a *WriterAdapter[T]) Create(ctx context.Context, m T) (*T, error) {
+	ri := a.client.Resource(a.gvr).Namespace(ComputeNamespace(m))
+
+	uobj, err := a.tToUnstructured(m)
+	if err != nil {
+		a.logger.ErrorContext(ctx, "conversion to k8s object failed", "resource", a.gvr.Resource, "error", err)
+		return nil, fmt.Errorf("%w: failed to convert %s to k8s object: %w", model.ErrValidation, a.gvr.Resource, err)
+	}
+
+	ures, err := ri.Create(ctx, uobj, metav1.CreateOptions{})
+	if err != nil {
+		a.logger.ErrorContext(ctx, "failed to create resource", "name", m.GetName(), "resource", a.gvr.Resource, "error", err)
+
+		var errModel error
+		switch {
+		case kerrs.IsNotFound(err): // occurs when the namespace of the resource does not exist
+			errModel = model.ErrNotFound
+		case kerrs.IsAlreadyExists(err): // occurs when the resource with the same name already exists
+			errModel = model.ErrAlreadyExists
+		case kerrs.IsInvalid(err): // occurs when the resource is semantically invalid
+			errModel = model.ErrValidation
+		default:
+			errModel = model.ErrUnavailable
+		}
+
+		return nil, fmt.Errorf("%w: failed to create %s '%s': %w", errModel, a.gvr.Resource, m.GetName(), err)
+	}
+
+	res, err := a.k8sToDomain(ures)
+	if err != nil {
+		a.logger.ErrorContext(ctx, "conversion from k8s object failed", "resource", a.gvr.Resource, "error", err)
+		return nil, fmt.Errorf("%w: failed to convert %s from k8s object: %w", model.ErrValidation, a.gvr.Resource, err)
+	}
+
+	return &res, nil
+}
+
+// Update implements the port.WriterRepo interface.
+func (a *WriterAdapter[T]) Update(ctx context.Context, m T) (*T, error) {
+	ri := a.client.Resource(a.gvr).Namespace(ComputeNamespace(m))
+
+	uobj, err := a.tToUnstructured(m)
+	if err != nil {
+		a.logger.ErrorContext(ctx, "conversion from T to unstructured failed", "resource", a.gvr.Resource, "error", err)
+		return nil, fmt.Errorf("%w: failed to convert %s to unstructured: %w", model.ErrValidation, a.gvr.Resource, err)
+	}
+
+	ures, err := ri.Update(ctx, uobj, metav1.UpdateOptions{})
+	if err != nil {
+		a.logger.ErrorContext(ctx, "failed to update resource", "name", m.GetName(), "resource", a.gvr.Resource, "error", err)
+
+		var errModel error
+		switch {
+		case kerrs.IsNotFound(err): // occurs when the resource to be updated does not exist or its namespace does not exist
+			errModel = model.ErrNotFound
+		case kerrs.IsConflict(err): // occurs when there is a resource version conflict during update
+			errModel = model.ErrConflict
+		case kerrs.IsInvalid(err): // occurs when the updated resource is semantically invalid
+			errModel = model.ErrValidation
+		default:
+			errModel = model.ErrUnavailable
+		}
+
+		return nil, fmt.Errorf("%w: failed to update %s '%s': %w", errModel, a.gvr.Resource, m.GetName(), err)
```

Comment:
Should we propagate the k8s error here? Since we've converted it already and logged it.

Reply:
actually that should be a handler concern probably